src/code.jl

```diff
@@ -1112,4 +1115,109 @@ function cse!(x::ForLoop, state::CSEState)
     return ForLoop(x.itervar, cse!(x.range, state), apply_cse(cse!(x.body, new_state), new_state))
 end
 
+"""
+    OptimizationRule(name, detector, transformer, priority)
+
+Defines an optimization rule with:
+- `name`: A string identifier for the optimization.
+- `detector`: A function that detects patterns in the IR. 
+- `transformer`: A function that transforms the IR based on detected patterns, and returns updated IR
+- `priority`: Integer priority (higher = applied first)
+
+The detector function should implement the signature
+
+```julia
+detector(expr::Code.Let, state::Code.CSEState) -> Union{Nothing, Vector{<:AbstractMatched}}
+```
+
+Likewise, the transformer function should implement the signature
+
+```julia
+transformer(expr::Code.Let, match_data::Union{Nothing, Vector{<:AbstractMatched}}, state::Code.CSEState) -> Code.Let
+```
+"""
+struct OptimizationRule{N, D, T, P}
+    name::N
+    detector::D
+    transformer::T
+    priority::P
+end
+
+abstract type AbstractMatched end
+
+function bank(dic, key, value)
```

Comment:
All of these utility functions should be documented. It's not clear what the arguments are or what the function is intended to do.